### PR TITLE
Extend pending state

### LINKED ISSUE #39: Admins should be able to "activate" users
- There should be an overview of all pending users
- Clicking an "approve" button should send an email welcoming them and asking them to confirm their email.
- A new user state should be added, saying the user's email is not yet activated. Maybe `pending` should be split into `pending_activation` and `pending_email`?

### PR DESCRIPTION
**Pull request by mcbloch**

### CODE REVIEW

**Comment by mcbloch** [comment]
image
image

**Comment by mcbloch** [comment]
Fixes #39

File: src/controllers/users_controller.rs [hunk 1]
```diff
@@ -153,6 +155,25 @@ pub fn set_admin(
...
 
+#[post("/users/<id>/approve")]
+pub fn set_approved(
+	id: i32,
+	_session: AdminSession,
+	conn: DbConn,
+) -> Result<impl Responder<'static>> {
+	let mut user = User::find(id, &conn)?;
+
+	if user.state == UserState::PendingApproval {
```
Comment (rien) [inline src/controllers/users_controller.rs:166]: Maybe an `else` returning a 422 error would be appropriate here.

File: src/models/user.rs [hunk 2]
```diff
@@ -15,15 +15,21 @@ use validator::{Validate, ValidationError};
...
 	Active,
 	Disabled,
 }
 
 impl fmt::Display for UserState {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
 		match *self {
-			UserState::Pending => write!(f, "Pending"),
+			UserState::PendingApproval => {
+				write!(f, "Pending for admin approval")
```
Comment (rien) [inline src/models/user.rs:28]:
```suggestion
				write!(f, "Admin approval pending")
```

File: src/models/user.rs [hunk 3]
```diff
@@ -15,15 +15,21 @@ use validator::{Validate, ValidationError};
...
+			},
+			UserState::PendingMailConfirmation => {
+				write!(f, "Pending for mail confirmation")
```
Comment (rien) [inline src/models/user.rs:31]:
```suggestion
				write!(f, "Email confirmation pending")
```

File: src/errors.rs [hunk 4]
```diff
@@ -110,8 +110,10 @@ pub type InternalResult<T> = std::result::Result<T, InternalError>;
 pub enum LoginError {
 	#[error("Username or password incorrect")]
 	UsernamePasswordError,
-	#[error("Account still pending")]
-	AccountPendingError,
+	#[error("Account still pending for admin approval")]
```
Comment (rien) [inline src/errors.rs:113]:
```suggestion
	#[error("Admin approval pending for this accountl")]
```

File: src/errors.rs [hunk 5]
```diff
@@ -110,8 +110,10 @@ pub type InternalResult<T> = std::result::Result<T, InternalError>;
...
+	AccountPendingApprovalError,
+	#[error("Account pending for mail approval")]
```
Comment (rien) [inline src/errors.rs:115]:
```suggestion
	#[error("Email confirmation pending for this account")]
```

File: src/models/user.rs [hunk 6]
```diff
@@ -288,8 +301,13 @@ impl User {
 			Ok(user) if !verify(password, &user.hashed_password) => {
 				Err(ZauthError::LoginError(LoginError::UsernamePasswordError))
 			},
-			Ok(user) if user.state == UserState::Pending => {
-				Err(ZauthError::LoginError(LoginError::AccountPendingError))
+			Ok(user) if user.state == UserState::PendingApproval => Err(
```
Comment (rien) [inline src/models/user.rs:304]: I think we want a `match user.state` here to make sure we handle all cases right in the future.

File: templates/users/index.html [hunk 7]
```diff
@@ -25,37 +34,86 @@ <h1>Users ({{ users.len() }})</h1>
...
-        <td>{{ user.last_login }}</td>
-        <td>{{ user.created_at }}</td>
+        <td>
+            {% match user.state %}
+                {% when UserState::Active %}
+                    {%- call colored_circle("#80A66C", user.state) -%}
+                {% when UserState::PendingApproval %}
+                    {%- call colored_circle("#F2A950", user.state) -%}
+                {% when UserState::PendingMailConfirmation %}
+                    {%- call colored_circle("yellow", user.state) -%}
```
Comment (rien) [inline templates/users/index.html:50]:
```suggestion
                    {%- call colored_circle("#ffff00", user.state) -%}
```

**Pull request merged**